# Hacker news data set
* hacker news is a popular site where technology related stories and posts are voted and commented upon similar to reddit.hacker newa is extermely popular in technology and start up circles, and the posts that make it to the top of hacker listings can get hundreds of thousands of visitors as a result.
* In this project we will work with a data set of subbmissions to popular techology site [hacker news.](https://news.ycombinator.com/)
* You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.
* We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. 

# introduction:
* first we will open the hacker_news.csv file and make it a list of lists and remove the header row from it.

In [1]:
from csv import reader
opened_file=open("hacker_news-copy1.csv")
read_file=reader(opened_file)
hn=list(read_file)
print(hn[0:3])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']]


In [2]:
header=hn[0]
hn=hn[1:]
print(hn[:3])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']]


## seperating the data set into 3 parts:
* the data set which we are given is divided into three parts :
1. ask: it stores all the post titles beginning with ASK HN.
2. show: it stores all the post titles beginning with SHOW HN.
3. other: it stores other files which dont have either ASK HN or SHOW HN in their titles.

In [3]:
ask=[]
show=[]
other=[]
for i in hn:
    title=i[1]
    title_l=title.lower()
    if title_l.startswith("ask hn")==True:
        ask.append(i)
    elif title_l.startswith("show hn")==True:
        show.append(i)
    else:
        other.append(i)
print("number of elements in ask: ",len(ask))
print("number of elements in show:  ",len(show))
print("number of elements in other: ",len(other))

number of elements in ask:  1744
number of elements in show:   1162
number of elements in other:  17194


* we will now print the first five rows of ask list and you can observe that all the titles in this list start with ASK HN.
* we will also going to print the first five rows of show list and all the titles start with SHOW HN.

In [4]:
for i in ask[:5]:
    print(i)
print('')
print('')
for j in show[:5]:
    print(j)

['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']
['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']
['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']
['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']
['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']


['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03']
['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']
['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', 

## Determining the average number of points in ask and show lists .
* to find the average number of points in ask and show lists we need to first initiate ask_c and show_c to 0
* loop through the ask list, number of points is in the fifth coloumn in ask list we need to get the element at index 3 in each row. dont forget to convert the value to integer type to perform addition to it.
* calculate points and assign it to variable ask_c .
* similarly do the same operations on show list and find the show_c
* after that calculate average points for both the lists.

In [5]:
ask_p=0
show_p=0
for i in ask:
    points_a=int(i[3])
    ask_p+=points_a
print("total number of points for ask posts:  ",ask_p)
for j in show:
    points_s=int(j[3])
    show_p+=points_s
print("total number of points for show posts: ",show_p)
print("average ask points: ",ask_p/len(ask))
print("average show points: ",show_p/len(show))

total number of points for ask posts:   26268
total number of points for show posts:  32019
average ask points:  15.061926605504587
average show points:  27.555077452667813


* now we will create 2 dictionaries which will store counts and points per hour
* first import datetime module as dt because we will need it to convert the created_at coloumn at index 6 in hour.
* then create an empty list named result list.which we will be a list of lists
* append the date and points to the list. here date and points form for each row form a list and stored into the bigger list res_show
* create 2 dictionaries count_by_hour and points_by_hour which will be a frequency table where the key is hour and the value is number of posts , number of points for that hour.

In [6]:
import datetime as dt
res_show=[]
for i in show:
    res_show.append([i[6],int(i[3])])
for i in res_show:
    time=i[0]
    dformat=("%m/%d/%Y %H:%M")
    hour=dt.datetime.strptime(time,dformat).strftime("%H")
    i[0]=hour

In [7]:
count_by_hour={}
points_by_hour={}
for i in res_show:
    hour=i[0]
    points=i[1]
    if hour in count_by_hour:
        count_by_hour[hour]+=1
        points_by_hour[hour]+=points
    else:
        count_by_hour[hour]=1
        points_by_hour[hour]=points
print(count_by_hour)
print("\n")
print("\n")
print(points_by_hour)

{'14': 86, '22': 46, '18': 61, '07': 26, '20': 60, '05': 19, '16': 93, '19': 55, '15': 78, '03': 27, '17': 93, '06': 16, '02': 30, '13': 99, '08': 34, '21': 47, '04': 26, '11': 44, '12': 61, '23': 36, '09': 30, '01': 28, '10': 36, '00': 31}




{'14': 2187, '22': 1856, '18': 2215, '07': 494, '20': 1819, '05': 104, '16': 2634, '19': 1702, '15': 2228, '03': 679, '17': 2521, '06': 375, '02': 340, '13': 2438, '08': 519, '21': 866, '04': 386, '11': 1480, '12': 2543, '23': 1526, '09': 553, '01': 700, '10': 681, '00': 1173}


In [8]:
average_points_by_hour={}
for i in points_by_hour:
    average_points_by_hour[i]=points_by_hour[i]/count_by_hour[i]
print(average_points_by_hour)

{'14': 25.430232558139537, '22': 40.34782608695652, '18': 36.31147540983606, '07': 19.0, '20': 30.316666666666666, '05': 5.473684210526316, '16': 28.322580645161292, '19': 30.945454545454545, '15': 28.564102564102566, '03': 25.14814814814815, '17': 27.107526881720432, '06': 23.4375, '02': 11.333333333333334, '13': 24.626262626262626, '08': 15.264705882352942, '21': 18.425531914893618, '04': 14.846153846153847, '11': 33.63636363636363, '12': 41.68852459016394, '23': 42.388888888888886, '09': 18.433333333333334, '01': 25.0, '10': 18.916666666666668, '00': 37.83870967741935}


* we will now create a list which store lists within it and each sublist is of 2 elements where the first element is hour and the second element is average points for that hour

In [9]:
avg_points=[]
for i in average_points_by_hour:
    avg_points.append([average_points_by_hour[i],int(i)])
print(avg_points)

[[25.430232558139537, 14], [40.34782608695652, 22], [36.31147540983606, 18], [19.0, 7], [30.316666666666666, 20], [5.473684210526316, 5], [28.322580645161292, 16], [30.945454545454545, 19], [28.564102564102566, 15], [25.14814814814815, 3], [27.107526881720432, 17], [23.4375, 6], [11.333333333333334, 2], [24.626262626262626, 13], [15.264705882352942, 8], [18.425531914893618, 21], [14.846153846153847, 4], [33.63636363636363, 11], [41.68852459016394, 12], [42.388888888888886, 23], [18.433333333333334, 9], [25.0, 1], [18.916666666666668, 10], [37.83870967741935, 0]]


# Sorting and printing values from a list of lists

In [10]:
sorted_avg_points=sorted(avg_points,reverse=True)
print("\n")
print("\n")
print(sorted_avg_points)





[[42.388888888888886, 23], [41.68852459016394, 12], [40.34782608695652, 22], [37.83870967741935, 0], [36.31147540983606, 18], [33.63636363636363, 11], [30.945454545454545, 19], [30.316666666666666, 20], [28.564102564102566, 15], [28.322580645161292, 16], [27.107526881720432, 17], [25.430232558139537, 14], [25.14814814814815, 3], [25.0, 1], [24.626262626262626, 13], [23.4375, 6], [19.0, 7], [18.916666666666668, 10], [18.433333333333334, 9], [18.425531914893618, 21], [15.264705882352942, 8], [14.846153846153847, 4], [11.333333333333334, 2], [5.473684210526316, 5]]


In [11]:
for i in sorted_avg_points:
    print("{}:00 : {:.2f} average points per post".format(i[1],i[0]))

23:00 : 42.39 average points per post
12:00 : 41.69 average points per post
22:00 : 40.35 average points per post
0:00 : 37.84 average points per post
18:00 : 36.31 average points per post
11:00 : 33.64 average points per post
19:00 : 30.95 average points per post
20:00 : 30.32 average points per post
15:00 : 28.56 average points per post
16:00 : 28.32 average points per post
17:00 : 27.11 average points per post
14:00 : 25.43 average points per post
3:00 : 25.15 average points per post
1:00 : 25.00 average points per post
13:00 : 24.63 average points per post
6:00 : 23.44 average points per post
7:00 : 19.00 average points per post
10:00 : 18.92 average points per post
9:00 : 18.43 average points per post
21:00 : 18.43 average points per post
8:00 : 15.26 average points per post
4:00 : 14.85 average points per post
2:00 : 11.33 average points per post
5:00 : 5.47 average points per post


# conclusion:
* clearly we can see that 23:00 hour of the day tht is 11:00 pm has the maximum of 42.39 average points per post.
* so you will probably receive more points if you post SHOW HN type at 11:00 pm.